### ISSUE TITLE
Support yml extension

### ISSUE THREAD

**Issue by h2non** [Masterminds/glide#130]
Both `yml` and `yaml` extensions are valid for YAML files. This change should be trivial, but helpful.

**Comment by technosophos** [comment]
I'm not sure I understand the request. Where in particular where you looking for us to support something different?

If I recall correctly, the `-y` flag will take _any_ extension and still attempt to parse it as YAML.

The `-f` flag lets you set whatever filename you want, and will write to it regardless of the extension.

The default filename is `Glide.yaml`. I'm not sure it's a good idea to have more than one default filename, which is essentially what we'd get if we allowed `Glide.yaml` and `Glide.yml`. That would lead to unpredictable results if both a `Glide.yaml` and a `Glide.yml` file were present. But it's certainly possible today to do `glide -y Glide.yml up`.

(At one point, were were talking about naming the default file `Glidefile`. We added `.yaml` simply so that editors could provide syntax and highlighting.)

Thanks!

**Comment by h2non** [comment]
IMHO this should be transparently supporting during the glide YAML file discovery process.
If  `glide.yaml` doesn't exists, fallback looking for `glide.yml`. Is that simple. Specific flag is great but useless except you're using a file with another completely different name.

**Comment by mattfarina** [comment]
FYI, I'm looking at how we handle config files. Looking at a couple ways to change things up to make it easier.

**Comment by dolmen** [comment]
@h2non I don't think that multiple names would help. So far you have not told us your reason why you want this feature.

If you want to call your file `Glide.yml` (maybe because your text editor prefers it), either fix your text editor or just make a symbolic link: `ln -s Glide.yaml Glide.yml`.

**Comment by h2non** [comment]
Things are simpler. Officially both `.yml` and `.yaml` file extensions are registered and supported for the YAML language. 

If you need an real example of an extended use case of the simplified file extension I'm talking about you can take a look to this repository and verify that Travis config file is called `.travis.yml`.

We're just talking here about a trivial change that helps users significantly and provides a convenient implementation based on YAML community standards.
 
Resolving one or other file is trivial, and I can only list benefits, such as: doesn't introduces any kind of constraint in terms of the utility behavior, makes the utility more useful and robust for end users and it's less error prone.

**Comment by mattfarina** [comment]
@h2non There are a couple things going on here. First, with the yaml file...

1. Isn't it that case that a majority of the time Glide is creating it for you?
2. What happens when you have a `glide.yaml` and `glide.yml` file?

The delay on acting on this is because we're looking at moving to a json spec. It seems folks would prefer json (and don't worry we'll fall back to the `glide.yaml`). I'm working on [github.com/mattfarina/pkg](https://github.com/mattfarina/pkg). When I incorporate that handling I'll rewrite that code. This is my next big change to Glide.

Any thoughts on json v. yaml?

**Comment by h2non** [comment]
> Isn't it that case that a majority of the time Glide is creating it for you?

In my personal case I created the file manually. Not a big deal anyway, but users should be able to pick how they want to create the manifest file.

> What happens when you have a glide.yaml and glide.yml file?

Simple: for back-draws compatibility and legacy behavior coherency, lookup first for `glide.yaml`, if it doesn't exists in the current (or even ancestors) directory, fallback looking for `glide.yml` following the same lookup algorithm. If no one exists, fail with an error. 

Regarding to the Go `pkg` spec, I think it's a good initiative, so well done, but IMHO wouldn't tend to use JSON to declare the manifest file, mostly because it's a machine interchange focused language, instead of a human interchange language. Also I think it's too much error prone for human editing (I had a lot of bad experiences with package.json/bower.json files when developers edit them manually and forget to close a quote or add a comma).

It's true that JSOn is faster and simpler to parse and serialize and has native support in modern languages, including Go, but I'd prefer picking another more declarative human focused languages such as YAML or TOML.

**Comment by technosophos** [comment]
In my view, we're not talking about supporting `.yml` as a general extension. We do that already. You can use it with `-y` and `-f` flags, etc. All we're really talking about in this issue is a very simple question:

- Is there a good reason to support TWO default filenames (`Glide.yaml` and `Glide.yml`) instead of one?

There are three big issues surrounding the proposal to support two default names: 

1. Two is definitely an increase in complexity in both finding and generating files. It's not a horrible increase in complexity, though. 
2. It also makes for a less-clear user experience ("What if we have both?"). But again, not horrible.
3. Finally, is it common to see programs like this that have multiple default files? I'm not aware that Bundler, Composer, Gems, Berkshelf, Mix, and other common ones do this. Am I wrong about that?

We need to table the JSON vs. YAML topic on this issue @mattfarina. It's going to confuse things. Though I suppose that if we're claiming that it's too confusing to have two extensions for the same file, it's an obvious leap that it's too confusing to have two entirely different formats.

**Comment by h2non** [comment]
Honestly, I can't see any of the problems you exposed like a significant problems which has more constraints that benefits.

The YAML extension ambiguity is far away from a new discussion, and most utilities that works with YAML-based files and worth to respect supports both extensions, mostly by a simple main reason: back-draws compatibility. 

At this point the discussion seems to me far away from obvious things, however to give you another rationale, if Glide uses TOML, XML or JSON instead of YAML, we don't have anything to discuss in this matter, but you've picked YAML, and YAML has two official extensions. Everyone supporting YAML has to support both extensions. That's the reality.

The most extended extension for YAML is `.yml`, and you can verify its adoption looking Travis, Ruby on Rails, CodeClimate... just to mention a few.

I agree about moving the discussion outside of the issue, but no one said supporting to different formats.

**Comment by technosophos** [comment]
Wait... you're getting off topic again, @h2non. We _do_ support `.yml`. The only relevant question is whether we support ONE default filename (`Glide.yaml`) or TWO default filenames (`Glide.yaml` and `Glide.yml`). You can already use `Glide.yml` as a file name using switches like `-f` and `-y`.

At this point, I'd welcome any PRs that support both filenames.

**Comment by h2non** [comment]
Roger that! Take a look to the PR!
